Remove root from snapshot

joshuagl opened the pull request:
**Fixes issue #**: #933 

**Description of the changes being introduced by the pull request**:
Per the updated spec in theupdateframework/specification#31 `root.json` should no longer be listed in `snapshot.json`.

* Remove the listing of `root.json` from `snapshot.json` and update the tests
* Update the docs/METADATA.md entry on `snapshot.json` 
* Update docstrings and code comments that refer to the order of operations for top-level roles to reflect the workflow in the latest spec
* Modify the test repository generator to remove a warning, update a comment and pass relative paths to the target files in `add_target()` calls
* Regenerate the repository_data for the `snapshot.json` change

**Please verify and check that the pull request fulfills the following
requirements**:

- [ ] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [ ] Docs have been added for the bug fix or new feature

joshuagl committed:
Remove root from snapshot

In PR #40 aginst the specification "root.json" has been removed from
the meta dictionary in "snapshot.json".

Update generate_snapshot_metadata() to no longer add an entry for
root.json to root.json

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Update docs/METADATA.md to reflect recent spec change

The specification was updated in PR #40 to remove root.json from
snapshot.json

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Re-generate repository and client test metadata

Re-generate metadata to adopt the change that root.json is no longer
listed in snapshot.json

```
 # Remove repository and client data
cd tests/repository_data && rm -rf repository client
 # Generate metadata
python generate.py
 # Duplicate metadata files
cp -r client/test_repository1 client/test_repository2
 # Recover non-signed file
git checkout client/map.json
```

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Update tests for removal of root.json from snapshot.json

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl committed:
Update test repository data generator

* Fix the path referenced in the Purpose
* Change add_target() calls to pass file paths relative to targets dir

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented on tests/repository_data/generate.py:
```diff
@@ -119,12 +119,12 @@
-repository.targets.add_target(target2_filepath)
+repository.targets.add_target(os.path.basename(target1_filepath), file_permissions)
+repository.targets.add_target(os.path.basename(target2_filepath))
 
```
As @lukpueh's hint in #933 was to make changes like this in order to re-generate the test repository data I wondered why we aren't making this change to the `generate.py` script, rather than hacking it each time?

I will dig into this and try to better understand it.

joshuagl replied:
As @lukpueh's hint in #933 was to make changes like this in order to re-generate the test repository data I wondered why we aren't making this change to the `generate.py` script, rather than hacking it each time?

I will dig into this and try to better understand it.

joshuagl commented on tests/test_updater.py:
```diff
@@ -886,7 +886,8 @@ def test_3__update_metadata_if_changed(self):
     self.repository_updater._update_metadata_if_changed('targets')
-    self.repository_updater._update_metadata_if_changed('root')
+    # TODO: should this still work?
+    # self.repository_updater._update_metadata_if_changed('root')
```
I need to read through the spec and code to figure out whether this should still work, it throws:
```
Traceback (most recent call last):
  File "/home/joshuagl/Projects/tuf/tests/test_updater.py", line 890, in test_3__update_metadata_if_changed
    self.repository_updater._update_metadata_if_changed('root')
  File "/home/joshuagl/Projects/tuf/tuf/client/updater.py", line 1915, in _update_metadata_if_changed
    ['meta'][metadata_filename]
KeyError: 'root.json'
```

since the removal of `root.json` from `snapshot.json`, which I believe makes sense, but I have yet to verify (hence draft PR).

joshuagl replied:
Removal of this test is correct, the `root` metadata is no longer referenced by `snapshot`, hence the traceback above. Furthermore, the root role is updated separately and verified based on the existing root metadata, following the detailed client workflow logic in the spec, in the `_update_root_metadata()` function.

joshuagl reviewed (commented):
The PR is marked as draft as I'm sharing before I've finished the changes.

joshuagl committed:
Update docstrings and comments for the Updater object

The workflow for downloading metadata for top-level roles has changed.

Root is now updated and verified by stepping through a chain of trust
based on the currently available root metadata. For that reason
root.json is no longer needed in snapshot and has been dropped from
there per theupdateframework/specification#40

Update docstrings and comments in the Updater object to reflect the
correct flow of metadata updates:
root (if necessary) -> timestamp -> snapshot -> targets

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
I believe this PR is ready for review

mnm678 commented on tests/repository_data/client/test_repository1/metadata/current/targets.json:
```diff
@@ -40,7 +40,7 @@
    "file1.txt": {
     "custom": {
-     "file_permissions": "0644"
+     "file_permissions": "0664"
```
Why did the write bit get added to these?

trishankatdatadog replied:
Good catch, Marina! Also, do we discuss anywhere why we set file permissions here at all? Is it show what the "custom" attribute can do? If so, we really ought to document it somewhere.

joshuagl replied:
> Why did the write bit get added to these?

Nice catch!

664 is the default file permissions for user-created files on Fedora (my host OS) and `generate.py` does not explicitly set permissions on the target files at creation time. Presumably this differs from wherever the original files were generated.

I can edit the patch to drop this change?

joshuagl replied:
Default file permissions on Ubuntu and Debian are 644 (umask `0022` vs umask `0002` on Fedora), which I believe would account for the discrepancy. I'll edit the patch to drop this change.

joshuagl replied:
I added an extra patch efae583 to set the permissions on `file1.txt` so that we don't generate different data on different Linux host OS.

trishankatdatadog commented on tests/repository_data/client/test_repository1/metadata/current/targets.json:
```diff
@@ -28,7 +28,7 @@
      "name": "role1",
      "paths": [
-      "file3.txt"
+      "/file3.txt"
```
I thought we removed adding a leading path separator to targets paths?

joshuagl replied:
Good eye, this was caused by changing:
`repository.targets.delegate('role1', [delegation_public], [target3_filepath])`
to
`repository.targets.delegate('role1', [delegation_public], [], list_of_targets=[os.path.basename(target3_filepath)])`

I made that change because I was trying to do away with the incorrect warning from `generate.py` that:
> 'file3.txt' is not located in the repository's targets directory: '/home/joshuagl/Projects/tuf/tests/repository_data/repository/targets'

I misunderstood the documentation for `delegate()` on first read, therefore I think it makes sense to undo the change to the `delegate()` call in the generate tool and regenerate the repository_data for this PR.

Then separate from this PR I can create a new Issue to clarify the docs and remove the incorrect warning in `repository_tool.targets.delegate()`

joshuagl replied:
In fact, I think the warning isn't even incorrect. At the time that we call `delegate()` in `generate.py` the path being delegated is not yet listed in the targets metadata because we are delegating the path before we have added a target file (with `add_target()`) at that path.

This was all misunderstanding on my part, thanks for the diligent review @trishankatdatadog

trishankatdatadog commented:
Other than @mnm678 and my comments, this PR LGTM!

joshuagl committed:
Improve determinism of test repository generator

One of the created target files has its file permissions encoded in the
targets metadata via the custom attribute of the add_target() function.
On Linux-based OS the umask value of the environment the script is run
in can result in different octal permissions for the created file, i.e.
on Fedora the default umask is 0002 (default permissions 664) whereas
on Debian/Ubuntu the default umask is 0022 (default permissions 644).

Explicitly chown 'file1' to octal permissions 644 so that the generated
data has the same custom attributes for targets regardless of which
Linux host they are generated on.

Signed-off-by: Joshua Lock <jlock@vmware.com>

trishankatdatadog commented:
@joshuagl Yeah, unfortunately the tuf logger is very noisy, and sometimes misleading...

lukpueh commented on docs/METADATA.md:
```diff
@@ -64,7 +64,7 @@ of delegated Targets metadata and [example](https://raw.githubusercontent.com/th
 Signed by: Snapshot role.
 
-The snapshot.json metadata file lists hashes and sizes of all metadata files other than timestamp.json. This file ensures that clients will see a consistent view of the files on the repository. That is, metadata files (and thus target file) that existed on the repository at different times cannot be combined and presented to clients by an attacker.
+The snapshot.json metadata file lists the version, and optionally the file hashes and sizes, of all metadata files other than root.json and timestamp.json. This file ensures that clients will see a consistent view of the files on the repository. That is, metadata files (and thus target file) that existed on the repository at different times cannot be combined and presented to clients by an attack ...
```
Minor nit: The *"all metadata files other"* wording is absolutely correct, but wouldn't it be more helpful to tell the reader what files to include, i.e. *"top-level targets metadata and all delegated targets metadata"* (instead of what files not to include)?
What do others think? This is definitely not a blocker for this PR. 

I proposed a similar change in the spec (see https://github.com/theupdateframework/specification/commit/8dc300f3d24298d3ccc3d154cd6221ed3f22843b and the newer https://github.com/theupdateframework/specification/commit/a5848e758a7aab27cb7143ecec287251a4f6d382)

joshuagl replied:
That would be much more helpful indeed. I've force-pushed an update to this patch which changes this to:
> The snapshot.json metadata file lists the version, and optionally the file hashes and sizes, of the top-level targets metadata and all delegated targets metadata.

lukpueh commented on tests/repository_data/generate.py:
```diff
@@ -119,12 +124,12 @@
+repository.targets('role1').add_target(os.path.basename(target3_filepath))
 repository.targets('role1').load_signing_key(delegation_private)
 
 repository.targets('role1').delegate('role2', [delegation_public], [])
```
Thanks for taking the time to actually fixing this. It makes more sense than adding a dirty patch to a commit message. :)

lukpueh commented on tuf/client/updater.py:
```diff
@@ -1836,21 +1831,23 @@ def _update_metadata_if_changed(self, metadata_role,
       it is updated in refresh() by calling _update_metadata('timestamp').
+      The 'root' role is always updated first and verified based on the trusted
+      root metadata file the client already has a copy of; it is updated in
+      refresh() by calling _update_root_metadata().
```
Thanks for making the effort of updating the docstrings. If I understand this function correctly, then it should only be called for "referenced metadata" (On an unrelated side-note: I think the parameter should be called `referencing_metadata`). 

This also means that that the function should no longer be called for root metadata as it is not referenced by any other metadata.  Would you mind removing the `elif metadata_role == 'root':`-logic and `TODO: If 'metadata_role' is root `-comment in the function body?

joshuagl replied:
I've pushed an additional patch to address the latter part of this comment

lukpueh reviewed (changes requested):
Excellent work and exemplary commit discipline, @joshuagl! I have one tiny request (see inline comment) otherwise it looks great to me.

By the way, I also tested your updates with the old metadata and tests still pass, which is great!

joshuagl committed:
Remove redundant code branch in Updater

Remove logic for handling of root metadata in _update_metadata_if_changed()
as root metadata is no longer fetched with this function, instead
_update_root_metadata() serves this purpose.

Additionally remove redundant mention of root metadata in a TODO comment.

Signed-off-by: Joshua Lock <jlock@vmware.com>

joshuagl commented:
Thank you for the diligent and exhaustive review, as always, @lukpueh ! Contributing to tuf is always a pleasure.

> By the way, I also tested your updates with the old metadata and tests still pass, which is great!

I should have thought to test that, happy to read it! 😄
